cpp/include/raft/neighbors/dataset.hpp

```diff
@@ -0,0 +1,296 @@
+  -> std::unique_ptr<strided_dataset<typename SrcT::value_type, typename SrcT::index_type>>
+{
+  using value_type       = typename SrcT::value_type;
+  using index_type       = typename SrcT::index_type;
+  using out_type         = strided_dataset<value_type, index_type>;
+  constexpr size_t kSize = sizeof(value_type);
+  uint32_t required_stride =
+    raft::round_up_safe<size_t>(src.extent(1) * kSize, align_bytes) / kSize;
+  return std::unique_ptr<out_type>{construct_strided_dataset(res, src, required_stride).release()};
+}
+
+template <typename DatasetT>
+auto upcast_dataset_ptr(std::unique_ptr<DatasetT>&& src)
+  -> std::unique_ptr<dataset<typename DatasetT::index_type>>
+{
+  using out_type = dataset<typename DatasetT::index_type>;
+  static_assert(std::is_base_of_v<out_type, DatasetT>, "The source must be a child of `dataset`");
+  return std::unique_ptr<out_type>{src.release()};
+}
+
+/** Parameters for VPQ compression. */
+struct vpq_params {
+  /**
+   * The bit length of the vector element after compression by PQ.
+   *
+   * Possible values: [4, 5, 6, 7, 8].
+   *
+   * Hint: the smaller the 'pq_bits', the smaller the index size and the better the search
+   * performance, but the lower the recall.
+   */
+  uint32_t pq_bits = 8;
+  /**
+   * The dimensionality of the vector after compression by PQ.
+   * When zero, an optimal value is selected using a heuristic.
+   *
+   * TODO: at the moment `dim` must be a multiple `pq_dim`.
+   */
+  uint32_t pq_dim = 0;
+  /**
+   * Vector Quantization (VQ) codebook size - number of "coarse cluster centers".
+   * When zero, an optimal value is selected using a heuristic.
+   */
+  uint32_t vq_n_centers = 0;
+  /** The number of iterations searching for kmeans centers (both VQ & PQ phases). */
+  uint32_t kmeans_n_iters = 25;
+  /**
+   * The fraction of data to use during iterative kmeans building (VQ phase).
+   * When zero, an optimal value is selected using a heuristic.
+   */
+  double vq_kmeans_trainset_fraction = 0;
+  /**
+   * The fraction of data to use during iterative kmeans building (PQ phase).
+   * When zero, an optimal value is selected using a heuristic.
+   */
+  double pq_kmeans_trainset_fraction = 0;
+};
+
+/**
+ * @brief VPQ compressed dataset.
+ *
+ * Twice quantized data:
+ *
+ *   1. Vector Quantization
+ *   2. Product Quantization of residuals
+ *
+ * @tparam MathT the type of elements in the codebooks
+ * @tparam IdxT type of the vector indices (represent dataset.extent(0))
+ *
+ */
+template <typename MathT, typename IdxT>
+struct vpq_dataset : public dataset<IdxT> {
+  /** Vector Quantization codebook - "coarse cluster centers". */
+  device_matrix<MathT, uint32_t, row_major> vq_code_book;
+  /** Product Quantization codebook - "fine cluster centers".  */
+  device_matrix<MathT, uint32_t, row_major> pq_code_book;
+  /** Compressed dataset.  */
+  device_matrix<uint8_t, IdxT, row_major> data;
```

Comment:
The `IdxT` here is used for consistency with the other dataset types (mdarrays) that are already in CAGRA and IVF-* methods, thus I'd prefer not to change it here.
I like the idea of removing the `IdxT` type though, it would simplify the whole hierarchy of dataset types (it's  only used for extents in this file). I'd suggest we could revise this (for all ANN algorithms together) in a follow-on PR or  in cuVS.

Reply:
I do not see how the "consistency" argument applies here. In CAGRA we have `int64_t` as mdarray extent type. https://github.com/rapidsai/raft/blob/2d714cbfda6de5a29858b423e4c8b37a75675014/cpp/include/raft/neighbors/cagra_types.hpp#L353-L356

If we use the mdspan index operator, then the offsets would be calculated using the extent type. For large arrays like `data`, I think it is important to use 64 bit types.

Reply:
This hasn't changed in the PR, because I instantiate the `neighbors::dataset` type with `int64_t` as well. That is, in the current state, we only ever instantiate the dataset with one extents type.
At the same time, we're using `IdxT` in IVF-* methods. I think there's no harm to keep the `IdxT` in the dataset (aiming to converge IVF-* and CAGRA methods one day) as long as it does not change the behavior of CAGRA.